File: 1250-longest-common-subsequence/longest-common-subsequence.cpp
class Solution {
public:
    int dp[1001][1001];
    int longestCommonSubsequence(string s1, string s2) {
        int n = s1.length(), m = s2.length();

        dp[0][0] = s1[0] == s2[0];
        for (int i = 1; i < n; i++) {
            dp[i][0] = (s1[i] == s2[0]);
            dp[i][0] = max(dp[i-1][0], dp[i][0]);
        }
        for (int i = 1; i < m; i++) {
            dp[0][i] = (s1[0] == s2[i]);
            dp[0][i] = max(dp[0][i-1], dp[0][i]);
        }

        for(int i = 1; i < n; i++)
            for(int j = 1; j < m; j++) {
                dp[i][j] = max(dp[i-1][j], dp[i][j-1]);
                if (s1[i] == s2[j]) dp[i][j] = max(dp[i][j], 1 + dp[i-1][j-1]);
            }

        return dp[n-1][m-1];
    }
};

File: 1250-longest-common-subsequence/longest-common-subsequence.py
class Solution(object):
    def longestCommonSubsequence(self, text1, text2):
        """
        :type text1: str
        :type text2: str
        :rtype: int
        """
        db = [[0 for i in range(len(text1))] for j in range(len(text2))]

        db[0][0] = 1 if text1[0] == text2[0] else 0 

        for j in range(1, len(text1)):
            if text1[j] == text2[0]:
                db[0][j] = 1
            else:
                db[0][j] = db[0][j-1]

        for i in range(1, len(text2)):
            if text1[0] == text2[i]:
                db[i][0] = 1
            else:
                db[i][0] = db[i-1][0]

        for j in range(1, len(text1)):
            for i in range(1, len(text2)):
                if text1[j] == text2[i]:
                    db[i][j] = 1+ db[i-1][j-1]
                else:
                    db[i][j] = max(db[i-1][j], db[i][j-1])
        return db[len(text2)-1][len(text1)-1]
        print(db)